e2e/nomostest/gitproviders/repository.go

```diff
@@ -51,193 +52,223 @@ const (
-
 	// RemoteRepoName is the name of the remote repository.
 	// It is the same as Name for the testing git-server.
 	// For other git providers, it appends a UUID to Name for uniqueness.
 	RemoteRepoName string
-
 	// GitProvider is the provider that hosts the Git repositories.
-	GitProvider gitproviders.GitProvider
-
+	GitProvider GitProvider
 	// Scheme used for encoding and decoding objects.
 	Scheme *runtime.Scheme
-
 	// Logger for methods to use.
 	Logger *testlogger.TestLogger
+
+	yamlSerializer *jserializer.Serializer
 }
 
 // NewRepository creates a remote repo on the git provider.
 // Locally, it writes the repository to `tmpdir`/repos/`name`.
-//
-// The repo name is in the format of <NAMESPACE>/<NAME> of RootSync|RepoSync.
-func NewRepository(nt *NT, repoType RepoType, nn types.NamespacedName, sourceFormat filesystem.SourceFormat) *Repository {
-	nt.T.Helper()
-
-	namespacedName := nn.String()
+// Name is the <NAMESPACE>/<NAME> of the RootSync|RepoSync.
+func NewRepository(
+	repoType RepoType,
+	syncNN types.NamespacedName,
+	sourceFormat filesystem.SourceFormat,
+	scheme *runtime.Scheme,
+	logger *testlogger.TestLogger,
+	provider GitProvider,
+	tmpDir string,
+	privateKeyPath string,
+) *Repository {
+	namespacedName := syncNN.String()
 	safetyName := fmt.Sprintf("safety-%s", strings.ReplaceAll(namespacedName, "/", "-"))
-
-	localDir := filepath.Join(nt.TmpDir, "repos", namespacedName)
-
-	g := &Repository{
-		Root:                  localDir,
+	return &Repository{
+		Name:                  namespacedName,
+		Root:                  filepath.Join(tmpDir, "repos", namespacedName),
 		Format:                sourceFormat,
-		T:                     nt.T,
 		Type:                  repoType,
 		SafetyNSName:          safetyName,
 		SafetyNSPath:          fmt.Sprintf("acme/namespaces/%s/ns.yaml", safetyName),
 		SafetyClusterRoleName: safetyName,
 		SafetyClusterRolePath: fmt.Sprintf("acme/cluster/cluster-role-%s.yaml", safetyName),
-		Scheme:                nt.Scheme,
-		Logger:                nt.Logger,
-		GitProvider:           nt.GitProvider,
+		Scheme:                scheme,
+		Logger:                logger,
+		GitProvider:           provider,
+		PrivateKeyPath:        privateKeyPath,
+		yamlSerializer:        jserializer.NewYAMLSerializer(jserializer.DefaultMetaFactory, scheme, scheme),
 	}
+}
 
-	repoName, err := nt.GitProvider.CreateRepository(namespacedName)
-	// Add the repo to nt.RemoteRepositories immediately after it is created to reuse the repo.
-	nt.RemoteRepositories[nn] = g
+// Create the remote repository using the GitProvider, and
+// create the local repository with an initial commit.
+func (g *Repository) Create() error {
+	repoName, err := g.GitProvider.CreateRepository(g.Name)
 	if err != nil {
-		nt.T.Fatal(err)
+		return errors.Wrapf(err, "creating repo: %s", g.Name)
 	}
 	g.RemoteRepoName = repoName
 
-	g.init(nt.gitPrivateKeyPath)
-	g.initialCommit(sourceFormat)
+	// TODO: Move this to the caller or GitProvider
```

Comment:
Maybe we could have an NT wrapper method that calls this and sets the entry in `RemoteRepositories`

EDIT: looks like you addressed this, so I think this comment can be removed

Reply:
Yeah, I moved it into ResetRepository, which is the sole caller of NewRepository now.
I'll remove the TODO.